### PR TITLE
dns: disable the use of search namespaces when using the c-ares resolver

### PR DESCRIPTION
**Pull request by suniltheta**
Signed-off-by: Sunil Narasimhamurthy <sunnrs@amazon.com>

Commit Message: Modify API to support configuration of ARES flags to disable the use of search namespaces when using the c-ares resolver and reconcile DNS options in a single message
Additional Description:
Risk Level: low
Testing: unit test
Docs Changes: yes
Release Notes: updated new feature and deprecated list
Platform Specific Features:
Issues: #12432
Deprecated: field 'use_tcp_for_dns_lookups' in bootstrap, cluster & dynamic_forward_proxy xDS APIs

### CODE REVIEW

**Comment by suniltheta** [comment]
Hi, this is a draft PR to address the issue https://github.com/envoyproxy/envoy/issues/12432 . 
I want to get the opinion and suggestions about the DnsLookupOptions fields. Can I get the tips on how I can propagate this message throughout the code where the DNS resolvers are instantiated.

Also need consensus on this change where use_tcp_for_dns_lookups option will be deprecated and will be moved inside new protobuf message DnsLookupOptions.

> Note: I have not included the files from running the _tools/proto_format/proto_format.sh fix_, to keep the change set small. So CI might be failing.

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [inline api/envoy/config/core/v3/base.proto:449]: FWIW I think we are heading towards a model of pluggable DNS resolvers in the not too distant future. I wonder if the DNS lookup options should reflect that?

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (adisuissa) [inline api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:104]: Can you add to the comment a reference to the configuration that should be used (instead of the deprecated one)?

File: api/envoy/config/core/v3/base.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/base.proto:461]:
```suggestion
  google.protobuf.BoolValue no_default_search_domain = 2;
```

File: api/envoy/config/core/v3/base.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/base.proto:457]:
```suggestion
  // server startup. Apple's API only uses UDP for DNS resolution.
```

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this. I've added a few comments.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto:449]: Yes, good idea. I never thought of that. Thank you for your suggestion. I will keep this in mind and take care of it in upcoming commits.

File: api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto
Comment (suniltheta) [reply api/envoy/extensions/common/dynamic_forward_proxy/v3/dns_cache.proto:107]: Added the comment.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto]: Took care of this manually. thanks for noticing.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto]: Took care of this manually. thanks for noticing.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (adisuissa) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:254]: I think there's a wrong link here. Should probably be:
`envoy_api_field_config.bootstrap.v3.Bootstrap.dns_lookup_options`.
Please fix here and in other places.

**Review by adisuissa — CHANGES_REQUESTED** [review]
Looks good.
Please fix link issues, as it blocks CI.

File: api/envoy/config/core/v3/base.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/base.proto:460]: Also I think this should be tagged with `[#not-implemented-hide:]`, because IIUC it is not implemented yet.

File: api/envoy/config/core/v3/base.proto
Comment (adisuissa) [inline api/envoy/config/core/v3/base.proto:457]: @htuch this new message replaces a deprecated field (`use_tcp_for_dns_lookups`). However, it also modifies the field type from `bool` to `google.protobuf.BoolValue`.
I'm assuming it is ok, as the other field is deprecated, just want your opinion on this.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (suniltheta) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:254]: Thanks for taking a look. I am working on these changes on the side. I will try to push rest of the changes by tomorrow. I will fix this :ref: link as well.

File: api/envoy/config/core/v3/base.proto
Comment (htuch) [reply api/envoy/config/core/v3/base.proto:457]: I think this is fine. What is the default value for this field? `BoolValue` is mostly useful when we anticipate wanting to change the default (in particular for security related fields).

Comment (adisuissa) [reply api/envoy/config/core/v3/base.proto:457]: The default value is protobuf's bool default value (`false`).
Actually, the name of the field `use_tcp_for_dns_lookups` implies it is meant to be "enabled". I'm not certain what's the added value of `BoolValue` in this case.

Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto:457]: So here is my approach to keep supporting the deprecated field "bool use_tcp_for_dns_lookups". I would love to get your opinions on them.

Even though the field "**_bool use_tcp_for_dns_lookups_**" is marked deprecated, we expect previous implementations of xDSv3 control plane to still set this value. So the logic will check if the new field "_google.protobuf.BoolValue use_tcp_for_dns_lookups_" is being set via _dns_lookup_options_ or not. The code will rely on **_has_**__use_tcp_for_dns_lookups()_ to check if this value is being programmed by the server. If this _dns_lookup_options_**_.has_use_tcp_for_dns_lookups()_** is false (which means value is not set) then the code will check if the deprecated field **_bool use_tcp_for_dns_lookups_** has the value **_true_** and uses it. This makes sure of the backward compatibility of the changes. 

Correct me if I am wrong here, by just using bool field type we will not be able to know if the config (true or false) is applied or not.

One might ask can we just do **_OR_** of **_bool use_tcp_for_dns_lookups_** and **_dns_lookup_options.use_tcp_for_dns_lookups().value()_**, like -> `config.dns_lookup_options.use_tcp_for_dns_lookups().value() || config.use_tcp_for_dns_lookups`. But as per the logic which is being worked upon now, the new field
**_google.protobuf.BoolValue dns_lookup_options.use_tcp_for_dns_lookups_**  will get preference and override any value set by bare **_bool use_tcp_for_dns_lookups_** value.

Comment (htuch) [reply api/envoy/config/core/v3/base.proto:457]: I think you can use the presence of the `dns_lookup_options` message alone to detect that we aren't dealing with legacy configuration..

Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto:453]: Thanks for the suggestions. Not using BoolValue.

File: api/envoy/config/core/v3/base.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/base.proto:456]: All the changes are committed.

File: include/envoy/event/dispatcher.h
Comment (adisuissa) [inline include/envoy/event/dispatcher.h:207]: Should be removed.

File: source/common/event/dispatcher_impl.cc
Comment (adisuissa) [inline source/common/event/dispatcher_impl.cc:176]: Note that in this case the new `no_default_search_domain` flag won't be used to initialize the AppleDnsResolver.

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc
Comment (adisuissa) [inline source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc:35]: Same as above.
Also not sure if the two pieces of code can be refactored, as they are essentially the same.

File: source/server/server.cc
Comment (adisuissa) [inline source/server/server.cc:538]: Same as above.
I think this can be refactored to avoid code dup.

File: source/common/upstream/cluster_factory_impl.cc
Comment (adisuissa) [inline source/common/upstream/cluster_factory_impl.cc:105]: Can you use something like `CopyFrom` here?
Otherwise, more fields will require manual copying.

**Review by adisuissa — COMMENTED** [review]
API looks good.
Left some drive-by comments on the non-api code.

File: source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc
Comment (suniltheta) [reply source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc]: Took care of this change. Need your thoughts here. In the updated changes the code looks like
```
    envoy::config::core::v3::DnsResolverOptions dns_resolver_options;
    if (cluster.has_dns_resolver_options()) {
      dns_resolver_options.CopyFrom(cluster.dns_resolver_options());
    } else {
      // <code comment>
      dns_resolver_options.set_use_tcp_for_dns_lookups(cluster.use_tcp_for_dns_lookups());
    }
```
technically we don't need both the `if` & `else` condition. We could just do with 
``` 
  envoy::config::core::v3::DnsResolverOptions dns_resolver_options;
  dns_resolver_options.CopyFrom(cluster.dns_resolver_options());
  if (!cluster.has_dns_resolver_options()) {
    dns_resolver_options.set_use_tcp_for_dns_lookups(cluster.use_tcp_for_dns_lookups());
  }
```

But I thought it is better if we make it explicit that if legacy control plane has not configured `dns_resolver_options` then we stick to just setting `dns_resolver_options.use_tcp...` based on deprecated field `use_tcp_for_dns_lookups`.

File: source/common/upstream/cluster_factory_impl.cc
Comment (suniltheta) [reply source/common/upstream/cluster_factory_impl.cc]: done. thanks.

File: source/server/server.cc
Comment (suniltheta) [reply source/server/server.cc:538]: done

**Comment by suniltheta** [comment]
> Waiting on #16294 to be merged before merging this.

**Comment by suniltheta** [comment]
> The job running on agent Azure Pipelines 17 ran longer than the maximum time of 180 minutes.
restarting the tests

**Comment by suniltheta** [comment]
Hi @htuch & @adisuissa, this PR is ready for review.
The CI is failing with the following reason below. I thought let me just get the review comments and address those comments so that subsequent commit can trigger a new CI run. Thank you for your time.
```
##[error]The agent did not connect within the alloted time of 45 minute(s).
,##[error]The request: 395683 was abandoned due to an infrastructure failure. Notification of assignment to an agent was never received.
```

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (htuch) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:263]: @yanjunxiang-google this is the proto message that DNS resolver extension should live in.

File: api/envoy/config/core/v3/resolver.proto
Comment (htuch) [inline api/envoy/config/core/v3/resolver.proto:26]: I think you should add a version history entry for this change.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/config/core/v3/resolver.proto
Comment (suniltheta) [reply api/envoy/config/core/v3/resolver.proto:26]: Done, thanks.

**Comment by suniltheta** [comment]
Hi Maintainers, this PR is ready for merge?
@htuch & @adisuissa have given first pass approval. Thank you for your time.
Tagging for attention @mattklein123, @htuch @dio, @lizan

**Comment by mattklein123** [comment]
cc @alyssawilk this is a case for the bot of a PR not getting a non-API reviewer. I will review.

**Review by mattklein123 — APPROVED** [review]
Thank you!

**Comment by alyssawilk** [comment]
Ugh, this was definitely missed with old and new versions of the script, though I also see cases in maintainer-oncall where the API bug is definitely caught. Can't debug this easily since it's merged but please let me know if you see other cases :-(

**Comment by alyssawilk** [comment]
on no wait, this was caught, I just searched for the wrong string.  So should be fixed forward as long as on-call goes through the list :-)

**Comment by suniltheta** [comment]
The incompatibility is between this PR and the PR at https://github.com/envoyproxy/envoy/pull/16294 
And anything that is incompatible is not across a release. So we should be good here?

**Comment by htuch** [comment]
Yep, we allow breaking API changes within a 14 day window, see https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md.

Good catch though!

**Comment by jpeach** [comment]
@htuch The `upstream_resolvers` field in `DnsFilterConfig` was present in Envoy 1.18, but absent in Envoy 1.19. AFAICT that makes this an incompatible change, and it's also difficult for a control plane to support both since the field number was re-used.

**Comment by suniltheta** [comment]
https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md
> API versions tagged vNalpha. Within an alpha major version, arbitrary breaking changes are allowed.

@jpeach That is true, but the API is still v3alpha. Hope that is allowed.

**Comment by jpeach** [comment]
> https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md
> 
> > API versions tagged vNalpha. Within an alpha major version, arbitrary breaking changes are allowed.
> 
> @jpeach That is true, but the API is still v3alpha. Hope that is allowed.

Is it possible to upgrade across the breaking change? i.e. I have to serve 1.18 and 1.19 clients  concurrently.

**Pull request merged**
